src/core/UniswapV4.sol

```diff
@@ -0,0 +1,770 @@
+    //// tokens or multiplex-out tokens (any credit will be swept back into Settler). To avoid doing
+    //// a linear scan each time a new token is encountered, the transient storage slot named by
+    //// that token stores the index (actually index * 32 + 32) of the token in the array. The
+    //// function `_take` is responsible for iterating over the list of tokens and withdrawing any
+    //// credit to the appropriate recipient.
+    ////
+    //// `state` exists to reduce stack pressure and to simplify and gas-optimize the process of
+    //// swapping. By keeping track of the sell token on each hop, we're able to compress the
+    //// representation of the fills required to satisfy the swap. Most often in a swap, the tokens
+    //// in adjacent fills are somewhat in common. By caching these tokens, we avoid having them
+    //// appear multiple times in the calldata. Additionally, this caching helps us avoid reading
+    //// the credit of each fill's sell token.
+
+    /// Because we have to ABIEncode the arguments to `.swap(...)` and copy the `hookData` from
+    /// calldata into memory, we save gas be deallocating at the end of this function.
+    function _swap(PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
+        private
+        DANGEROUS_freeMemory
+        returns (BalanceDelta)
+    {
+        return IPoolManager(_operator()).swap(key, params, hookData);
+    }
+
+    // the mandatory fields are
+    // 2 - sell bps
+    // 1 - pool key tokens case
+    // 3 - pool fee
+    // 3 - pool tick spacing
+    // 20 - pool hooks
+    // 3 - hook data length
+    uint256 private constant _HOP_LENGTH = 32;
+
+    /// To save stack and to simplify the following helper functions, we cache a bunch of state
+    /// about the swap. Note that `globalSellAmount` is practically unused when selling a FoT token.
+    struct State {
+        IERC20 globalSellToken;
+        uint256 globalSellAmount;
+        IERC20 sellToken;
+        uint256 sellAmount;
+        IERC20 buyToken;
+        uint256 buyAmount;
+    }
+
+    /// Decode a `PoolKey` from its packed representation in `bytes`. Returns the suffix of the
+    /// bytes that are not consumed in the decoding process. The first byte of `data` describes
+    /// which of the compact representations for the hop is used.
+    ///   0 -> sell and buy tokens remain unchanged from the previous fill (pure multiplex)
+    ///   1 -> sell token remains unchanged from the previous fill, buy token is read from `data` (diamond multiplex)
+    ///   2 -> sell token becomes the buy token from the previous fill, new buy token is read from `data` (multihop)
+    ///   3 -> both sell and buy token are read from `data`
+    ///
+    /// This function is also responsible for calling `_note`, which maintains the `notes` array and
+    /// the corresponding mapping in transient storage
+    function _getPoolKey(
```

Comment:
nit: This does a lot more than just decodes a pool key, also switching the entire token context (and modifying `State`). `_processNextHop`?

Reply:
You'll be horrified to know that the version of this function in #233 does _even more_. It's responsible for not only decoding the next hop from `data` and setting `state`, but also managing `notes`. Although it's much better abstracted and encapsulated. Perhaps this function does too many things? Maybe the logic from setting `zeroForOne` and after should be lifted back out into the loop in `unlockCallback` and this function should be named `_decodeHopTokens`?